File: tests/cpp/test_preflight.cpp
#include <gtest/gtest.h>

#include <string>
#include <vector>

#include "starling/preflight.hpp"
#include "starling/profile_capability.hpp"

using starling::preflight;
using starling::PreflightResult;
using starling::PreflightStatus;
using starling::ProfileCapability;

namespace {

ProfileCapability make_local_store() {
    return ProfileCapability{
        .profile_name = "local-store",
        .relational_backend = "seekdb",
        .vector_backend = "seekdb",
        .graph_backend = "ladybugdb",
        .c_plus_plus_core = true,
        .cross_partition_transaction = true,
        .transactional_outbox = true,
        .consumer_checkpoint = true,
        .tenant_isolation = "storage_enforced",
        .engram_per_record_key = true,
        .engram_refcount = true,
        .projection_index_supported = false,
        .dimension_versions_supported = false,
        .testing_helper_marker = true,
    };
}

}  // namespace

TEST(Preflight, FullyCapableLocalStoreReturnsReady) {
    auto cap = make_local_store();
    auto required = std::vector<std::string>{
        "transactional_outbox",
        "consumer_checkpoint",
        "engram_per_record_key",
        "c_plus_plus_core",
        "tenant_isolation_storage_enforced",
        "cross_partition_transaction",
    };
    PreflightResult result = preflight(cap, required);
    EXPECT_EQ(result.status, PreflightStatus::READY);
    EXPECT_TRUE(result.missing.empty());
}

TEST(Preflight, MissingTransactionalOutboxReturnsUnready) {
    auto cap = make_local_store();
    cap.transactional_outbox = false;
    PreflightResult result = preflight(cap, {"transactional_outbox"});
    EXPECT_EQ(result.status, PreflightStatus::UNREADY);
    ASSERT_EQ(result.missing.size(), 1u);
    EXPECT_EQ(result.missing[0], "transactional_outbox");
}

TEST(Preflight, AppFilterFailsStorageEnforcedRequirement) {
    auto cap = make_local_store();
    cap.tenant_isolation = "app_filter";
    PreflightResult result =
        preflight(cap, {"tenant_isolation_storage_enforced"});
    EXPECT_EQ(result.status, PreflightStatus::UNREADY);
    ASSERT_EQ(result.missing.size(), 1u);
    EXPECT_EQ(result.missing[0], "tenant_isolation_storage_enforced");
}

TEST(Preflight, MissingCrossPartitionTransactionFailsLocalAtomic) {
    auto cap = make_local_store();
    cap.cross_partition_transaction = false;
    PreflightResult result = preflight(cap, {"cross_partition_transaction"});
    EXPECT_EQ(result.status, PreflightStatus::UNREADY);
    ASSERT_EQ(result.missing.size(), 1u);
    EXPECT_EQ(result.missing[0], "cross_partition_transaction");
}

TEST(Preflight, TestingHelperMarkerMissingFailsLocalStore) {
    auto cap = make_local_store();
    cap.testing_helper_marker = false;
    PreflightResult result = preflight(cap, {"testing_helper_marker"});
    EXPECT_EQ(result.status, PreflightStatus::UNREADY);
    ASSERT_EQ(result.missing.size(), 1u);
    EXPECT_EQ(result.missing[0], "testing_helper_marker");
}

TEST(Preflight, MultipleMissingPreservesOrder) {
    ProfileCapability cap;  // all defaults = false
    PreflightResult result = preflight(cap, {
        "transactional_outbox",
        "consumer_checkpoint",
        "engram_per_record_key",
    });
    EXPECT_EQ(result.status, PreflightStatus::UNREADY);
    ASSERT_EQ(result.missing.size(), 3u);
    EXPECT_EQ(result.missing[0], "transactional_outbox");
    EXPECT_EQ(result.missing[1], "consumer_checkpoint");
    EXPECT_EQ(result.missing[2], "engram_per_record_key");
}

File: tests/python/test_preflight.py
def _local_store(core):
    return core.ProfileCapability(
        profile_name="local-store",
        relational_backend="seekdb",
        vector_backend="seekdb",
        graph_backend="ladybugdb",
        c_plus_plus_core=True,
        cross_partition_transaction=True,
        transactional_outbox=True,
        consumer_checkpoint=True,
        tenant_isolation="storage_enforced",
        engram_per_record_key=True,
        engram_refcount=True,
        projection_index_supported=False,
        dimension_versions_supported=False,
        testing_helper_marker=True,
    )


def test_preflight_full_capability_returns_ready(core):
    cap = _local_store(core)
    result = core.preflight(cap, [
        "transactional_outbox",
        "consumer_checkpoint",
        "engram_per_record_key",
        "c_plus_plus_core",
        "tenant_isolation_storage_enforced",
        "cross_partition_transaction",
    ])
    assert result.status == core.PreflightStatus.READY
    assert result.missing == []


def test_preflight_missing_outbox_returns_unready(core):
    cap = _local_store(core)
    cap.transactional_outbox = False
    result = core.preflight(cap, ["transactional_outbox"])
    assert result.status == core.PreflightStatus.UNREADY
    assert result.missing == ["transactional_outbox"]


def test_preflight_app_filter_blocks_storage_enforced(core):
    cap = _local_store(core)
    cap.tenant_isolation = "app_filter"
    result = core.preflight(cap, ["tenant_isolation_storage_enforced"])
    assert result.status == core.PreflightStatus.UNREADY
    assert result.missing == ["tenant_isolation_storage_enforced"]


def test_preflight_unknown_capability_treated_as_missing(core):
    cap = _local_store(core)
    result = core.preflight(cap, ["totally_made_up_capability"])
    assert result.status == core.PreflightStatus.UNREADY
    assert result.missing == ["totally_made_up_capability"]
